migrate-the-clay-components-from-v2-to-v3.md

```diff
@@ -94,6 +94,59 @@ To get to the behavior of having a ClayLink with image, use the composition with
 
 -   Added ability to utilize context for passing spritemap down instead of having to pass the prop everywhere.
 
+## ClayModal
+
+Of more freedom to customize `Footer` but still requiring some rules to make everything work well, always use the `<ClayModal.FooterContainer />` component to align the left and right buttons.
+
+```diff
+<ClayModal
+-	body={<h1>{'Hello World!'}</h1>}
++	onClose={() => setVisible(false)}
+	visible={visible}
+-/>
++>
++	<h1>{'Hello World!'}</h1>
++</ClayModal>
+```
+
+### API Changes
+
+-  `body` deprecated in favor of utilizing `children` prop
+-  `footerButtons` removed in favor of utilizing `renderFooter` prop
+-  `onClose` added
+-  `visible` is the source of truth to open or close the modal, `show` method has been removed.
+-  `data` deprecated
+-  `defaultEventHandler` deprecated
+-  `elementClasses` renamed to `className`
+
+### Compositions
+
+To render the buttons to the footer use the utility and follow some rules to have no problem, in case you forget the component has issued errors in the console to remind you.
+
+-	Do not wrap `FooterContainer` with an HTML tag, use `Fragment` to do so.
+-	Do not use more than two `FooterContainer`. 
+
+```diff
+<ClayModal
+-	body={<h1>{'Hello World!'}</h1>}
++	onClose={() => setVisible(false)}
++	renderFooter={
++		<>
++			<ClayModal.FooterContainer>
++				<ClayButton>{'Cancel'}</ClayButton>
++			</ClayModal.FooterContainer>
++			<ClayModal.FooterContainer>
++				<ClayButton>{'Save'}</ClayButton>
++			</ClayModal.FooterContainer>
++		</>
```

Comment:
I don't think this is a very ergonomic API. We should be able to grab the children directly and wrap them inside the component without requiring the user to do it.

Reply:
I also think. The issue that we have a limitation on markup, we need to define two elements that define the left and right side of the footer, in this case the two `FooterContainer` elements here (Abstracts this information, but I'm not sure if it was the right choice or if I need give more context in our documentation...), so I can add more than one ClayButton inside those containers. Even if I can eliminate the use of `FooterContainer` I would not have been able to identify this.

Maybe I could do something like:
```jsx
<>
  <>
    <ClayButton>{'Save'}</ClayButton>
    <ClayButton>{'Cancel'}</ClayButton>
  </>
  <>
    <ClayButton>{'...'}</ClayButton>
    <ClayButton>{'...'}</ClayButton>
  </>
</>
```

We can use the Fragment to differentiate the sides, but I do not know if this would be a recommended standard, I did not get to see such a case.
> I have not tested this yet

Reply:
The thing about getting people to group using nested fragments is that is very implicit and probably not the friendliest API.

At `$previousJob` we had a `<LeftRight />` component for this, that expected two children. Was a bit nicer to use than another alternative, which was separate props for left and right:

```javascript
<Thing
   primary={<OkButton />}
   secondary={<CancelButton />}
/>

<Thing>
   <LeftRight>
     <CancelButton />
     <OkButton />
   </LeftRight>
</Thing>
```

Right now I can't remember whether it automatically flipped for RTL locales. It probably did.

Reply:
Hmm, it's a good solution too. But this case the `<LeftRight />` had a pre-defined markup? I say in all the application.

Maybe it makes more sense to go with the left and right props to simplify and remove the `<FooterContainer />` elements. Although we can use `<LeftRight />`, for those who wish to use more than two buttons can add fragments, makes sense?

Reply:
`<LeftRight />` was just something like:

1. Assert that `props.children` contains exactly two children.
2. Render (something like):

```
<div>
  <div class="left">
    {child1}
  </div>
  <div class="right">
  </div>
</div>
```

Reply:
(Updated because I accidentally submitted.)

Reply:
I've been thinking and adding `<Footer />` element with the props left and right and leaving the prop `footerContent` from ClayModal just by accepting a `React.ReactElement`.

```
<Footer
  left={<ClayButton />}
  right={(
    <>
      <ClayButton />
      <ClayButton />
    </>
  )}
/>

<ClayModal
  footerContent={<any>}
/>
```

Reply:
Only thing I would want to check with Lexicon is whether it might be better to use semantic names like "primary" and "secondary" instead of spatial ones like "left" and "right". For example, if we expect to flip the order when in a RTL locale, then the semantic names might be less confusing. Up to you though.

Reply:
Hm yeah, hey @victorvalle any thoughts on this?

Reply:
hey @wincent, I've been thinking a bit about our conversations from yesterday and maybe it makes more sense to give more "freedom", that's just a rough sketch but I want your opinion on that.

```jsx
<Modal size="lg" status="danger">
  <Modal.Header onClose={} spritemap={}>
    Title
  </Modal.Header>
  <Modal.Body url={...}>
    <h1>Body</h1>
  </Modal.Body>
  <Modal.Footer
    primary={}
    secondary={}
  />
</Modal>
```

Reply:
@wincent @matuzalemsteles ClayCSS has:
```
<div class="modal-footer">
  <div class="modal-item-first"></div>
  <div class="modal-item></div>
  <div class="modal-item-last"></div>
</div>
```

See http://pat270.github.io/lexicon/v2.10.0/content/modals/#modal-variants (modal success button). Modal Danger has a markup error.

Reply:
Thanks @ pat270 I had forgotten about this, does it make sense then to name the props as `first`, `middle` and `last`?

Reply:
@matuzalemsteles yes, I think `first`, `middle`, `last` makes sense.